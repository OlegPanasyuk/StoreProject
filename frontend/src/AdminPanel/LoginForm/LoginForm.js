import React, { Component } from 'react';
import { Modal, Button, Form, Overlay, Tooltip } from 'react-bootstrap';
import {Redirect} from 'react-router-dom';

//Redux 
import {connect} from 'react-redux';
import {
    userAuthorizedSuccess
} from '../../REDUX/adminPanel/actions/actionsLoginForm';

//for Rest requests
import rest from 'rest';
import pathPrefix from 'rest/interceptor/pathPrefix';
import errorCode from 'rest/interceptor/errorCode';
import mime from 'rest/interceptor/mime';

const client = rest.wrap(mime, { mime: 'application/json' })
    .wrap(errorCode, { code: 500 })
    .wrap(pathPrefix, { prefix: `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`});


export class LoginForm extends Component {
    constructor(props) {
        super(props);
        this.attachRef = target => this.setState({ target });
        this.hide = this.hide.bind(this);
        this.sendRequestOnAccess = this.sendRequestOnAccess.bind(this);
        this.emailRef = React.createRef();
        this.passRef = React.createRef();
        
        this.state = {
            target: null,
            show: true,
            showTooltip: false,
            messageToolTip: '',

        };
    }

    hide() {
        this.setState((state) => ({
            show: !state.show
        }));
        let path = this.props.match.path;
        let p = path.slice(0, path.indexOf('/login'));
        window.location.pathname = `${p}`;
    }

    sendRequestOnAccess() {
        client({
            path: '/login',
            method: 'POST',
            entity: {
                email: this.emailRef.current.value,
                password: this.passRef.current.value
            }, 
            headers: {
                checkrights: true
            }
        })
            .then(data => {
                if (data.entity.token) {
                    this.props.userAuthorizedSuccess(data.entity);
                } else {
                    this.setState({
                        showTooltip: true,
                        message: data.entity
                    });
                }
            })
            .catch( err=> {
                console.log(err);
            });
    }

    render() {
        let loginCont = ''; 
        if (this.props.user.token) {
            let path = this.props.match.path;
            let p = path.slice(0, path.indexOf('/login'));
            loginCont = (
                <Redirect to={p} />
            );
        } 
        let {target, showTooltip, message } = this.state;
        return (
            <React.Fragment>
                <Modal
                    show={this.state.show}
                    size="sm"
                    centered
                >
                    <Modal.Header>
                        <h4>LoginForm</h4>
                    </Modal.Header>
                    <Form>
                        <Modal.Body>
                            <Form.Group>
                                <Form.Control 
                                    type="email" 
                                    placeholder="Enter email" 
                                    ref={this.emailRef}
                                />
                            </Form.Group>
                            <Form.Group>
                                <Form.Control 
                                    type="password" 
                                    placeholder="Password" 
                                    ref={this.passRef}
                                />       
                            </Form.Group>
                        </Modal.Body>
                        <Modal.Footer>
                            <Button
                                ref={this.attachRef}
                                onClick={this.sendRequestOnAccess}>
                                Login
                            </Button>
                            <Button 
                                onClick={this.hide}
                                variant = 'light'
                            >
                                Close
                            </Button>
                            
                        </Modal.Footer>
                    </Form>
                </Modal>
                {loginCont}
                <Overlay target={target} show={showTooltip} placement="left">
                    {props => (
                        <Tooltip id="overlay-example" {...props} show={showTooltip.toString()}>
                            {message}
                        </Tooltip>
                    )}
                </Overlay>
            </React.Fragment>
        );
    }
}

const mapStoreToProps = (state) => {
    return ({
        user: state.adminPanel.user
    });
};


export default connect(mapStoreToProps, {
    userAuthorizedSuccess
})(LoginForm);