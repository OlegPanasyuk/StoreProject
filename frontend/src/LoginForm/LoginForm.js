import React, { Component } from 'react';
import { Form, Col, Button } from 'react-bootstrap';
import Link from '../Link/Link';
import PropsTypes from 'prop-types';
import md5 from 'md5';

//Redux
import { connect } from 'react-redux';
import {
    addErrorToState,
} from '../REDUX/actions/actionsErrors';

import rest from 'rest';
import pathPrefix from 'rest/interceptor/pathPrefix';
import errorCode from 'rest/interceptor/errorCode';
import mime from 'rest/interceptor/mime';

const client = rest.wrap(mime, { mime: 'application/json' })
    .wrap(errorCode, { code: 500 })
    .wrap(pathPrefix, { prefix: 'http://localhost:3300' });

export class LoginForm extends Component {
    constructor(props) {
        super(props);
        this.attachRef = target => { this.setState({ target }); };
        this.emailInput = React.createRef();
        this.passWordInput = React.createRef();
        this.sendLoginRequest = this.sendLoginRequest.bind(this);
        this.state = {
            target: null,
            message: '',
            show: false
        };
    }

    sendLoginRequest(e) {
        e.preventDefault();
        const objToRequest = {
            email: this.emailInput.current.value,
            password: this.passWordInput.current.value
        };
        client({ method: 'POST', path: 'login', entity: objToRequest })
            .then(data => {
                //console.log(data);
                let storage = window.localStorage;

                if (data.status.code === 200) {
                    storage.setItem('Authorization', data.entity.token);
                    this.props.handleSetStateInApp(data.entity);
                } else if (data.status.code === 401) {
                    const d = new Date();
                    this.props.addErrorToState({
                        id: md5(`${'Notification from LoginForm'}${d.valueOf()}`),
                        level: 'Warning',
                        message: data.entity
                    });
                }

            }).catch((err) => {
                throw new Error('error in request', err);
            });
    }

    render() {
        let handleConverStatusUser = this.props.handleConverStatusUser;
        let userState = this.props.userState;
        return (
            <Form className='ml-auto'>
                <Form.Row className='d-flex align-items-center justify-content-end'>
                    <Form.Group as={Col} className="col mb-0" controlId="formGridEmail">
                        <Form.Control
                            size="sm"
                            type="email"
                            placeholder="Enter email"
                            ref={this.emailInput}
                            defaultValue={(userState.email) ? userState.email : ''}
                        />
                    </Form.Group>
                    <Form.Group as={Col} className="col mb-0" controlId="formGridPassword">
                        <Form.Control size="sm" type="password" placeholder="Password" ref={this.passWordInput} />
                    </Form.Group>
                    <Col className='d-flex align-items-center'>
                        <Button 
                            ref={this.attachRef}
                            size="sm" 
                            variant="secondary" 
                            className="" 
                            onClick={this.sendLoginRequest}
                        >
                            Login
                        </Button>
                        
                        <Link
                            text='Registration'
                            href='#'
                            className="d-block p-3"
                            ConverStatusUser={handleConverStatusUser}>
                        </Link>
                    </Col>
                </Form.Row>
            </Form>
        );
    }
}

LoginForm.propTypes = {
    userState: PropsTypes.object,
    handleConverStatusUser: PropsTypes.func,
    handleSetStateInApp: PropsTypes.func,
    addErrorToState: PropsTypes.func
};

export default connect(null, {
    addErrorToState
})(LoginForm);