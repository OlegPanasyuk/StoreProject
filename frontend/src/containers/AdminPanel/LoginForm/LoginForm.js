import React, { Component } from 'react';
import {
    Modal, Button, Form, Overlay, Tooltip
} from 'react-bootstrap';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';

// Redux
import { connect } from 'react-redux';

// for Rest requests
import rest from 'rest';
import pathPrefix from 'rest/interceptor/pathPrefix';
import errorCode from 'rest/interceptor/errorCode';
import mime from 'rest/interceptor/mime';
import {
    userAuthorizedSuccess
} from '../../../actions/adminPanel/actions/actionsLoginForm';

const client = rest.wrap(mime, { mime: 'application/json' })
    .wrap(errorCode, { code: 500 })
    .wrap(pathPrefix, { prefix: `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}` });


function handle(e) {
    if (e.keyCode === 13) {
        document.getElementById('buttonToLogin').click();
    }
}

export class LoginForm extends Component {
    constructor(props) {
        super(props);
        this.attachRef = target => this.setState({ target });
        this.hide = this.hide.bind(this);
        this.sendRequestOnAccess = this.sendRequestOnAccess.bind(this);
        this.emailRef = React.createRef();
        this.passRef = React.createRef();

        this.state = {
            target: null,
            show: true,
            showTooltip: false,
            emailValid: {
                isValid: false,
                isInValid: false
            }
        };
    }

    hide() {
        this.setState(state => ({
            show: !state.show
        }));
        const { match } = this.props;
        const { path } = match;
        const p = path.slice(0, path.indexOf('/login'));
        window.location.pathname = `${p}`;
    }

    preValid() {
        let answ = true;
        const regEmail = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/i;
        if (!regEmail.test(this.emailRef.current.value)) {
            answ = false;
            this.setState({
                emailValid: {
                    isValid: false,
                    isInValid: true
                }
            });
        } else {
            answ = true;
            this.setState({
                emailValid: {
                    isValid: true,
                    isInValid: false
                }
            });
        }
        return answ;
    }

    sendRequestOnAccess() {
        const { userAuthorizedSuccess } = this.props;
        if (this.preValid()) {
            client({
                path: '/login',
                method: 'POST',
                entity: {
                    email: this.emailRef.current.value,
                    password: this.passRef.current.value
                },
                headers: {
                    checkrights: true
                }
            })
                .then((data) => {
                    if (data.entity.token) {
                        window.onkeydown = null;
                        userAuthorizedSuccess(data.entity);
                    } else {
                        this.setState({
                            showTooltip: true,
                            message: data.entity
                        });
                    }
                })
                .catch((err) => {
                    this.setState({
                        showTooltip: true,
                        message: err.toString()
                    });
                });
        }
    }

    render() {
        const { user, match, show } = this.props;
        const { emailValid } = this.state;
        let loginCont = '';
        if (user.token) {
            const { path } = match;
            const p = path.slice(0, path.indexOf('/login'));
            loginCont = (
                <Redirect to={p} />
            );
        }
        const { target, showTooltip, message } = this.state;
        return (
            <React.Fragment>
                <Modal
                    show={show}
                    size='sm'
                    onHide={this.hide}
                    onEntering={() => {
                        window.onkeydown = handle;
                    }}
                    onExit={() => {
                        window.onkeydown = null;
                    }}
                    keyboard
                    centered
                >
                    <Modal.Header>
                        <h4>LoginForm</h4>
                    </Modal.Header>
                    <Form>
                        <Modal.Body>
                            <Form.Group>
                                <Form.Control
                                    type='email'
                                    placeholder='Enter email'
                                    ref={this.emailRef}
                                    isValid={emailValid.isValid}
                                    isInvalid={emailValid.isInValid}
                                />
                                <Form.Control.Feedback>Looks good!</Form.Control.Feedback>
                                <Form.Control.Feedback type='invalid'>Incorrect email</Form.Control.Feedback>
                            </Form.Group>
                            <Form.Group>
                                <Form.Control
                                    type='password'
                                    placeholder='Password'
                                    ref={this.passRef}
                                />
                            </Form.Group>
                        </Modal.Body>
                        <Modal.Footer>
                            <Button
                                ref={this.attachRef}
                                onClick={this.sendRequestOnAccess}
                                id='buttonToLogin'
                            >
                                Login
                            </Button>
                            <Button
                                onClick={this.hide}
                                variant='light'
                            >
                                Close
                            </Button>

                        </Modal.Footer>
                    </Form>
                </Modal>
                {loginCont}
                <Overlay target={target} show={showTooltip} placement='left'>
                    {props => (
                        <Tooltip id='overlay-example' {...props} show={showTooltip.toString()}>
                            {message}
                        </Tooltip>
                    )}
                </Overlay>
            </React.Fragment>
        );
    }
}

LoginForm.propTypes = {
    match: PropTypes.object,
    user: PropTypes.object,
    userAuthorizedSuccess: PropTypes.func,
    show: PropTypes.bool
};

LoginForm.defaultProps = {
    match: { url: '', path: '' },
    user: {},
    userAuthorizedSuccess: () => {},
    show: true
};

const mapStoreToProps = state => ({
    user: state.adminPanel_User
});


export default connect(mapStoreToProps, {
    userAuthorizedSuccess
})(LoginForm);
