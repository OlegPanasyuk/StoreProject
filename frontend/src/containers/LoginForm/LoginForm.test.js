import React from 'react';
import { shallow, configure } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import PropTypes from 'prop-types';
import { LoginForm } from './LoginForm';

configure({ adapter: new Adapter() });


function setupComponent(props = {
    handleConverStatusUser: () => {},
    handleSetStateInApp: () => {},
    userState: {
        email: 'test@test.com',
        role: 'Guest'
    }
}) {
    const component = shallow(<LoginForm {...props} />);
    return { component, props };
}


describe('LoginForm', () => {
    it('Component must be created', () => {
        const { component } = setupComponent();

        expect(LoginForm.propTypes.userState).toBe(PropTypes.object);

        expect(LoginForm.propTypes.handleConverStatusUser).toBe(PropTypes.func);

        expect(LoginForm.propTypes.handleSetStateInApp).toBe(PropTypes.func);

        expect(shallowToJson(component)).toMatchSnapshot();
    });
});
