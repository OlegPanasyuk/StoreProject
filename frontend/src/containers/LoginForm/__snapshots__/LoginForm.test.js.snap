// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component must be created 1`] = `
<Bootstrap(Modal)
  centered={true}
  keyboard={true}
  onEntering={[Function]}
  onExiting={[Function]}
  onHide={[Function]}
  show={true}
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <h4>
      LoginForm
    </h4>
  </ModalHeader>
  <ModalBody>
    <Form
      as="form"
      className="ml-auto"
      inline={false}
    >
      <FormGroup
        as="div"
        className=" mb-3"
        controlId="formGridEmail"
      >
        <FormControl
          as="input"
          defaultValue="test@test.com"
          isInvalid={false}
          isValid={false}
          placeholder="Enter email"
          size="sm"
          type="email"
        />
        <Feedback
          as="div"
          type="valid"
        />
        <Feedback
          as="div"
          type="invalid"
        />
      </FormGroup>
      <FormGroup
        as="div"
        className=" mb-3"
        controlId="formGridPassword"
      >
        <FormControl
          as="input"
          isInvalid={false}
          isValid={false}
          placeholder="Password"
          size="sm"
          type="password"
        />
        <Feedback
          as="div"
          type="valid"
        />
        <Feedback
          as="div"
          type="invalid"
        />
      </FormGroup>
      <div
        className="d-flex justify-content-end"
      >
        <Button
          active={false}
          disabled={false}
          id="buttonLoginForm"
          onClick={[Function]}
          size="sm"
          type="button"
          variant="secondary"
        >
          Login
        </Button>
        <NavLink
          className="d-block ml-3 mr-3"
          onClick={[Function]}
          to="/"
        >
          Registration
        </NavLink>
      </div>
    </Form>
  </ModalBody>
</Bootstrap(Modal)>
`;
